File: easysocket/SocketServer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using Newtonsoft.Json.Linq;

namespace easysocket
{
    public class SocketServer
    {
        public SocketServer(int port, string ipAdd = "127.0.0.1")
        {
            IPAddress ip = IPAddress.Parse(ipAdd);
            m_client_sockets = new Dictionary<string, ClientInfo>();
            m_serverSocket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
            m_serverSocket.Bind(new IPEndPoint(ip, port));  //绑定IP地址：端口
            m_serverSocket.Listen(10);    //设定最多10个排队连接请求
            Console.WriteLine("S> 启动监听端口{0}成功", m_serverSocket.LocalEndPoint.ToString());
            //通过Clientsoket发送数据
            m_listen_thread = new Thread(ListenClientConnect);
            m_listen_thread.Start();
        }

        public void StopListening()
        {
            if (m_listen_thread != null)
            {
                m_listen_thread.Abort();
                m_listen_thread = null;
            }
        }

        public ClientInfo GetClient(string name)
        {
            return m_client_sockets.ContainsKey(name)
                ? m_client_sockets[name]
                : null;
        }

        public ClientInfo[] GetAllClients()
        {
            return m_client_sockets.Values.ToArray();
        }

        private void ListenClientConnect()
        {
            while (true)
            {
                Socket clientSocket = m_serverSocket.Accept();
                clientSocket.Send(Encoding.UTF8.GetBytes("Server Say Hello"));
                (new Thread(ReceiveMessage)).Start(clientSocket);
            }
        }

        private void ReceiveMessage(object clientSocket)
        {
            Socket myClientSocket = (Socket)clientSocket;
            while (true)
            {
                try
                {
                    //通过clientSocket接收数据
                    int receiveNumber = myClientSocket.Receive(buffer);
                    string recvString = Encoding.UTF8.GetString(buffer, 0, receiveNumber);
                    if (recvString != "")
                    {
                        OnProcessJsonQuery(myClientSocket, recvString);
                    }
                    else
                    {
                        Console.WriteLine("S> 接收中断： 客户端可能主动断开了连接");
                        myClientSocket.Shutdown(SocketShutdown.Both);
                        myClientSocket.Close();
                        break;
                    }
                }
                catch (Exception err)
                {
                    Console.WriteLine("S> 接收中断： {0}:\n {1}", err.Message, err.StackTrace);
                    myClientSocket.Shutdown(SocketShutdown.Both);
                    myClientSocket.Close();
                    break;
                }
            }
        }

        public bool ClientLogin(ClientInfo client)
        {
            lock (m_client_sockets)
            {
                //TODO: 释放掉可能的旧链接
                m_client_sockets[client.Name] = client;
            }
            return true;
        }

        private bool OnProcessJsonQuery(Socket clientSocket, string recvString)
        {
            dynamic json = JObject.Parse(recvString);
            string clientName = json["name"].Value;
            try
            {
                //解析json并确定请求内容及请求者名称
                ClientInfo clientInfo = null;
                clientInfo = GetClient(clientName);
                if (clientInfo == null) { clientInfo = new ClientInfo(clientName, clientSocket, this); }
                clientInfo.Process(json);
                return true;
            }
            catch (Exception err)
            {
                Console.WriteLine("S> [{0}]处理失败： {1}: \n {2}", clientName, err.Message, err.StackTrace);
                clientSocket.Shutdown(SocketShutdown.Both);
                clientSocket.Close();
                return false;
            }
        }

        private void RemoveSocketFromClientSockets(Socket socket)
        {
            string removeKey = null;
            foreach(var pair in m_client_sockets)
            {
                if (pair.Value.Socket == socket)
                {
                    removeKey = pair.Key;
                    break;
                }
            }
            if (removeKey != null && m_client_sockets.ContainsKey(removeKey))
            {
                m_client_sockets.Remove(removeKey);
            }
        }

        public class ClientInfo
        {
            string m_name;
            SocketServer m_server;
            Socket m_socket;
            List<Message> m_send_data_wait;   //发送缓冲区

            public Socket Socket { get { return m_socket; } }

            public struct Message
            {
                public string From { get; }
                public string To { get; }
                public dynamic Data { get; }
                public Message(string from, string to, dynamic data)
                {
                    From = from;
                    To = to;
                    Data = data;
                }
            }

            public ClientInfo(string name, Socket socket, SocketServer server)
            {
                m_name = name;
                m_socket = socket;
                m_server = server;
                m_send_data_wait = new List<Message>();
            }
            public string Name { get { return m_name; } }
            public void SetData(Message msg)
            {
                lock (m_send_data_wait)
                {
                    m_send_data_wait.Add(msg);
                }
            }
            public void Process(dynamic queryJson)
            {
                JObject respondJson = new JObject();
                respondJson["name"] = "Server";
                respondJson["result"] = 1;
                string action = queryJson["action"].Value;
                if (action == "login")
                {
                    string message;
                    if (m_server.ClientLogin(this))
                    {
                        message = "登录成功";
                        respondJson["result"] = 0;
                    }
                    else
                    {
                        message = "登录失败";
                    }
                    respondJson["data"] = message;
                    Console.WriteLine("S> [{0}]{1}", Name, message);
                }
                else if (action == "send")
                {
                    string toClientName = queryJson["to"].Value;
                    dynamic sendData = queryJson["data"];
                    ClientInfo targetClient = m_server.GetClient(toClientName);
                    string message = "";
                    if (targetClient != null)
                    {
                        targetClient.SetData(new Message(Name, toClientName, sendData));
                        respondJson["result"] = 0;
                        message = "消息已转发";
                    }
                    else
                    {
                        message = "消息转发失败，目标客户端[" + toClientName + "]未发现";
                        respondJson["action"] = "error";
                    }
                    respondJson["data"] = message;
                    Console.WriteLine("S> {0}", message);
                }
                else if (action == "listen")
                {
                    while (m_send_data_wait.Count == 0) { Thread.Sleep(TimeSpan.FromSeconds(0.001)); }
                    var sendData = m_send_data_wait[0];
                    respondJson["from"] = sendData.From;
                    respondJson["data"] = sendData.Data;
                    respondJson["result"] = 0;
                    Console.WriteLine("S> 消息转发 [{0}] -> [{1}]", sendData.From, sendData.To);
                    m_send_data_wait.RemoveAt(0);
                }
                else if (action == "echo")
                {
                    respondJson["result"] = 0;
                    Console.WriteLine("S> 接收客户端[{0}]消息({1})", Name, queryJson.ToString());
                }
                else
                {
                    string message = "无法识别的请求：" + action;
                    respondJson["action"] = "error";
                    respondJson["data"] = message;
                    Console.WriteLine("S> 请求处理失败（" + respondJson.ToString() + "）");
                }
                m_socket.Send(Encoding.UTF8.GetBytes(respondJson.ToString()));
            }
        }

        private static byte[] buffer = new byte[100 * 1024];
        internal Socket m_serverSocket;
        internal Dictionary<string, ClientInfo> m_client_sockets;
        internal Thread m_listen_thread;
    }
}

File: easysocketPy/socket_server.py
import json
import socket
import logging
import traceback
import threading
import time

logging.basicConfig(format='%(asctime)s %(message)s', level=logging.DEBUG, filename='log.log')


class SocketServer(object):

    def __init__(self, ip_address, port):
        self.m_client_sockets = {}
        self.m_client_lock = threading.Lock()
        self.m_name = "服务器"
        self.m_server_socket = socket.socket()
        self.m_server_socket.bind((ip_address, port))
        """设定最多10个排队连接请求"""
        self.m_server_socket.listen(10)
        try:
            self.m_listen_thread = threading.Thread(target=self.listen_client_connect)
            self.m_listen_thread.start()
            self.m_listen_thread_run = 1
            self.normal_log("服务器监听线程开启: " + ip_address + ":" + str(port))
        except:
            self.m_listen_thread_run = 0
            self.error_log("开启监听线程失败")

    def error_log(self, title):
        logging.error("[" + self.m_name + "]> " + title + ": " + traceback.format_exc())

    def normal_log(self, text):
        logging.info("[" + self.m_name + "]> " + text)

    def stop_listening(self):
        if self.m_listen_thread is not None:
            self.m_listen_thread_run = 0
            self.m_listen_thread = None

    def get_client(self, name):
        return self.m_client_sockets[name] if name in self.m_client_sockets else None

    def get_all_clients(self):
        return self.m_client_sockets.values()

    def get_client_name(self, connection):
        for client_name in self.m_client_sockets.keys():
            if self.m_client_sockets[client_name].Socket == connection:
                return client_name
        return "未记录"

    def listen_client_connect(self):
        while True:
            connection, addr = self.m_server_socket.accept()
            respond_data = {"result": 0, "message": "连接成功"}
            connection.send(bytes(json.dumps(respond_data), encoding="utf8"))
            self.normal_log("接收连接请求：" + addr[0] + ":" + str(addr[1]))
            receive_thread = threading.Thread(target=self.receive_message, args=(connection,))
            receive_thread.start()

    def receive_message(self, connection: socket.socket):
        while True:
            try:
                receive_data = connection.recv(2048)
                if len(receive_data) == 0:
                    self.normal_log("接收中断： 客户端["+self.get_client_name(connection)+"]可能主动断开了连接")
                    self.remove_socket_from_client_sockets(connection)
                    connection.shutdown(socket.SHUT_RDWR)
                    connection.close()
                    break
                else:
                    if not self.on_process_json_query(connection, receive_data):
                        break
            except:
                self.error_log("接收中断")
                self.remove_socket_from_client_sockets(connection)
                break

    def client_login(self, client):
        self.m_client_lock.acquire()
        self.m_client_sockets[client.Name] = client
        self.m_client_lock.release()
        return True

    def on_process_json_query(self, client_socket, receive_data):
        try:
            data = json.loads(str(receive_data, encoding="utf-8"))
            client_name = data["name"] if "name" in data else ""
            if client_name != "":
                client_info = self.get_client(client_name)
                if client_info is None:
                    client_info = ClientInfo(client_name, client_socket, self)
                client_info.process(data)
                return True
            else:
                return False
        except:
            self.error_log("处理失败: " + str(receive_data, encoding="utf-8"))
            self.remove_socket_from_client_sockets(client_socket)
            client_socket.shutdown(socket.SHUT_RDWR)
            client_socket.close()
            return False

    def remove_socket_from_client_sockets(self, client_socket):
        remove_key = None
        for key in self.m_client_sockets:
            if self.m_client_sockets[key].Socket == client_socket:
                remove_key = key
                break
        if remove_key is not None and remove_key in self.m_client_sockets:
            self.m_client_sockets.pop(remove_key, None)
            self.normal_log("清理客户端数据: " + remove_key)


class Message(object):

    def __init__(self, from_client, to_client, data):
        self.m_from_client = from_client
        self.m_to_client = to_client
        self.m_data = data

    @property
    def From(self):
        return self.m_from_client

    @property
    def To(self):
        return self.m_to_client

    @property
    def Data(self):
        return self.m_data


class ClientInfo(object):

    def __init__(self, name, client_socket, server: SocketServer):
        self.m_name = name
        self.m_socket = client_socket
        self.m_server = server
        self.m_send_data_wait = []
        self.m_send_data_lock = threading.Lock()
        self.m_on_callbacks = {}
        self.regist_callback("login", self.on_login)
        self.regist_callback("send", self.on_send)
        self.regist_callback("listen", self.on_listen)
        self.regist_callback("echo", self.on_echo)

    def regist_callback(self, action, callback):
        self.m_on_callbacks[action] = callback

    def error_log(self, title):
        logging.error("[" + self.m_name + "]> " + title + ": " + traceback.format_exc())

    def normal_log(self, text):
        logging.info("[" + self.m_name + "]> " + text)

    @property
    def Name(self):
        return self.m_name

    @property
    def Socket(self):
        return self.m_socket

    def set_data(self, msg):
        self.m_send_data_lock.acquire()
        self.m_send_data_wait.append(msg)
        self.m_send_data_lock.release()

    def on_login(self, query_json, respond_json):
        if self.m_server.client_login(self):
            message = "登陆成功"
            respond_json["result"] = 0
        else:
            message = "登录失败"
        respond_json["data"] = message
        self.normal_log(message)
        return respond_json

    def on_send(self, query_json, respond_json):
        to_client_name = query_json["to"]
        send_data = query_json["data"]
        target_client = self.m_server.get_client(to_client_name)
        if target_client is not None:
            target_client.set_data(Message(self.Name, to_client_name, send_data))
            respond_json["result"] = 0
            message = "消息已转发"
        else:
            message = "消息转发失败，目标客户端[" + to_client_name + "]未发现"
            respond_json["action"] = "error"
        respond_json["data"] = message
        self.normal_log(message)
        return respond_json

    def on_listen(self, query_json, respond_json):
        while (len(self.m_send_data_wait) == 0):
            time.sleep(0.001)
        send_data = self.m_send_data_wait[0]
        respond_json["from"] = send_data.From
        respond_json["data"] = send_data.Data
        respond_json["result"] = 0
        message = "消息转发 [{0}] -> [{1}]"
        message = message.format(send_data.From, send_data.To)
        self.normal_log(message)
        self.m_send_data_lock.acquire()
        self.m_send_data_wait.pop(0)
        self.m_send_data_lock.release()
        return respond_json

    def on_echo(self, query_json, respond_json):
        respond_json["result"] = 0
        message = "接收客户端[{0}]消息({1})"
        message = message.format(self.Name, json.dumps(query_json))
        self.normal_log(message)
        return respond_json

    def process(self, query_json):
        respond_json = {"name": "Server", "result": 1}
        action = query_json["action"]
        if action in self.m_on_callbacks:
            respond_json = self.m_on_callbacks[action](query_json, respond_json)
        else:
            message = "无法识别的请求：" + action
            respond_json["action"] = "error"
            respond_json["data"] = message
            self.normal_log("请求处理失败(" + json.dumps(respond_json) + ")")
        self.m_socket.send(bytes(json.dumps(respond_json), encoding='utf8'))
